### ISSUE TITLE
4 Sequence Encoding Blocks You Must Know Besides RNN/LSTM in Tensorflow · Han Xiao Tech Blog - Deep Learning, Tensorflow, Machine Learning and more!

### ISSUE THREAD

**Issue by hanxiao** [hanxiao/hanxiao.github.io#10]
https://hanxiao.github.io/2018/06/25/4-Encoding-Blocks-You-Need-to-Know-Besides-LSTM-RNN-in-Tensorflow/

**Comment by hanxiao** [comment]
revision #1: add causality to CNN encode block

**Comment by romanovzky** [comment]
Amazing read! It's refreshing seeing someone discussing alternative sequence enconders!

Just a question, do you think these could be unsupervisedly train in autoencoders?

Cheers!

**Comment by hanxiao** [comment]
@romanovzky Yes it can be used in autoencoder settings.

**Comment by romanovzky** [comment]
And how do you often deal with different lengths? Do you always pad the short ones to a minimum length, or are there other solutions that can deal with varying lengths? I've always felt uncomfortable about the padding solution, although I know I shouldn't.

**Comment by hanxiao** [comment]
@romanovzky you should pad the sequences when a batch contains sequences with different lengths. It is the *correct* thing to do. But also remember to always mask out the computation op on those paddings, to ensure the correctness of your algorithm. The masking operation can be subtracting a large number before you output `logits`, or it can be setting corresponding positions to zero before you output `probabilities`. You have to keep those paddings in mind whenever you design an upper layer.

I also understand some people don't do padding seriously when you browse their code on Github, they sometimes mask and sometimes not. They may also achieve some good performance, but that doesn't mean their implementations are fully correct. DNN often has so many parameters and freedoms that even an imperfect implementation can yield good enough solution.

**Comment by switchfootsid** [comment]
Beautiful post!
